### ISSUE TITLE
Never inferred as Unknown in contravariant generic

### ISSUE THREAD

**Issue by mezuzza** [microsoft/pyright#4188]
**Describe the bug**
When using a contravariant type variable, a `Never` type argument is instead inferred as `Unknown`

**To Reproduce**
Run pyright on the code below

**Expected behavior**
`test3`'s return type should be `Foo[Never]` not `Foo[Unknown]`

**Screenshots or Code**
```python
from typing import Generic, NoReturn, TypeVar

from typing_extensions import Never, reveal_type

T = TypeVar("T", contravariant=True)


class A:
  pass


class B(A):
  pass


class Foo(Generic[T]):
  def bar(self, x: T):
    del x


class Bar(Foo[A]):
  pass


class Baz(Foo[B]):
  pass


def test(b: bool) -> Foo[B]:
  if b:
    return Bar()
  else:
    return Baz()


def test2(b: bool) -> Foo[A]:
  if b:
    return Bar()
  else:
    return Baz()  # Type error


def test3(b: bool) -> Foo[Never]:
  if b:
    return Bar()
  else:
    return Baz()


def test4(b: bool) -> Foo[NoReturn]:
  if b:
    return Bar()
  else:
    return Baz()


reveal_type(test)  # (b: bool) -> Foo[B]
reveal_type(test2)  # (b: bool) -> Foo[A]
reveal_type(test3)  # (b: bool) -> Foo[Unknown]
reveal_type(test4)  # (b: bool) -> Foo[NoReturn]
```

**VS Code extension or command-line**
Command line

**Additional context**
```sh
» pyright --version
pyright 1.1.279
» python --version
Python 3.10.8
```

**Comment by mezuzza** [comment]
Please feel free to correct my understanding of `Never` or contravariance is wrong. But my understanding is that `Foo` should be contravariant with respect to T as it only takes `T`s as input.

`Never` should be the bottom type so `Foo[Never]` should be a super type of all `Foo`. However, it seems I'm unable to express that.

**Comment by erictraut** [comment]
You are correct. The behavior you're seeing is left over from long before `Never` became supported officially in the type system. Pyright has internally used `Never` for a long time for type narrowing, but prior to it becoming an officially-supported type, I wanted to prevent it from appearing in type arguments for inferred return types. There was some old code that always translated `Never` type arguments to `Unknown`. I've removed that old code.

This will be addressed in the next release.

**Comment by mezuzza** [comment]
You guys are fast. 

Thanks for taking care of it.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.280, which I just published. It will also be included in a future version of pylance.
